### ISSUE TITLE
Crashes Without View

### ISSUE THREAD

**Issue by dodekeract** [hyperapp/hyperapp#318]
Believe it or not, one instance of hyperapp in one of my projects isn't using any view whatsoever. So here I am submitting a bug report to a frontend library that it crashes when no view is given...

## Broken Code [▶️](https://codepen.io/anon/pen/aydqVg?editors=0010)

```javascript
app({
    state: 'Example',
    events: { loaded: () => console.info('Loaded!') }
})
```

**Comment by jbucaran** [comment]
@dodekeract Two things to consider here:

1) Is this a user bug or 
2) do we want to support apps without a view?

To me the answer is (1) and this is the kind of thing I hope to catch using a dev build of the project someday. If the answer was (2) then we could simply add a default view of () => "".

Now, I don't see the point of an app without a view. You want a mixin instead.

**Comment by dodekeract** [comment]
@jbucaran I don't want a mixin instead. Thanks for assuming I don't know what I'm doing, but that app is running inside an iframe to take advantage of CORS to protect user privacy. It doesn't show anything and only communicates with an API. However, since it shares a lot of code with other iframes running in the same tab I need to take advantage of the rest of hyperapp's architecture (state, actions, events) to re-use code.

> If the answer was (2) then we could simply add a default view of () => "".
Why do we have to render anything at all? Neither `root` nor `view` is specified so the most sane thing is to render nothing.

**Comment by jbucaran** [comment]
@dodekeract What should hyperapp do in this case then?

**Comment by jbucaran** [comment]
I am not sure I want to provide you a empty view by default when you can simply do it like this:

```
view: () => ""
```

It seems that if this is really your use case, you have more

**Comment by dodekeract** [comment]
@jbucaran I just think that hyperapp shouldn't attempt to render if there is no view. It should just not attempt to render anything if there is no function to call.

Wouldn't `""` still render a text node?

**Comment by jbucaran** [comment]
>  Wouldn't "" still render a text node?

Correct.

>  It should just not attempt to render anything if there is no function to call.

This would complicate `render` and `schedule` and I don't know if introducing this complexity is any worth it.

**Comment by dodekeract** [comment]
I have one app per frame. That's the same as having multiple web applications. The only place where it makes a difference regarding hyperapp is for my reasoning to use it even though I don't actually need any view.

**Comment by jbucaran** [comment]
> @dodekeract I have one app per frame. 

That's interesting. I guess this is not exactly what I was thinking above.

---

I am not exactly sure how to proceed here. 

99% of the time you use a view. You may not need a view at this moment inside your iframe app, but you may change your mind later and decide to add it.

**Comment by lukejacksonn** [comment]
If you remove view and render from a hyperapp, then you are essentially left with a state management library and event bus.. which, if worked as expected, could be useful I guess?

I've not tried it out but presumably, this could be made to work by adjusting the update function https://github.com/hyperapp/hyperapp/blob/master/src/app.js#L37 to look something like:

```js
function update(withState) {
  if (withState != null) {
    state = merge(state, emit("update", withState))
    if(view) repaint(state)
  }
}
```

- What are the pros/cons of such an implementation?
- Does it disrupt `size`, `perf` or the `api`?

**Comment by jbucaran** [comment]
@lukejacksonn We call repaint in two places. So, it is either adding another `if (view)` or inside `render`, but then we end up scheduling anyway, just not calling `view`.

Another implementation option is to make the view `()=>""` by default.

**Comment by jbucaran** [comment]
> Does it disrupt size, perf or the api?

Size and perf for an extremely little, I guess.

**Comment by dodekeract** [comment]
@jbucaran I'm not even sure if it actually affects the performance at all. V8 might be able to optimise it to not even run the check.

**Comment by jbucaran** [comment]
@dodekeract Can you share some of the code or a **non-working** example that helps visualizing the architecture you are working with.

And, if you were using a different library, like React, what do you think you'd do?

**Comment by dodekeract** [comment]
I can't share any real code, but here is some pseudocode that basically shows what's going on:

```javascript
app({
    state: {},
    mixins: [SomeSharedMixin],
    events: {
        loaded: (state, actions) => window.addEventListener('message', actions.onMessage)
    }
})
```

Where `SomeSharedMixin` handles the `onMessage` action.

As you can see, I don't need a view in *this instance*, but since `SomeSharedMixin` also is used for communication in all other instances of hyperapp (running in a different frame though) it doesn't make sense to re-write that.

**Comment by jbucaran** [comment]
@dodekeract Good news! I'll add support for view-less apps in the next release. 👍

I'll implement this simply by not calling the `render` function. An app without a view will not fire a `render` event either.

**Comment by dodekeract** [comment]
@jbucaran Good, it's easy to workaround, but doesn't make sense to crash needlessly.

**Comment by selfup** [comment]
@lukejacksonn Seems that a hyperapp instance without a view will be very similar to EventHub in Vue but (in my eyes) will be simpler. Something like this I imagine:

_Codepen example: https://codepen.io/selfup/pen/jLMRjO

```javascript
/** @jsx h */
const { h, app } = hyperapp;

const HYPERHUB = app({
  state: {},
  actions: {
    hello: (state, actions, { id }) => { console.log('hello', id) },
  },
  events: {
    dispatch: (state, actions, { action, data }) => {
      return actions[action]
        ? actions[action](data)
        : null;
    },
  },
  view: (state, actions) => <span></span>,
});

const dispatch = (action, data) => HYPERHUB(
  'dispatch',
  { action, data },
);

dispatch('hello', { id: 42 });
```

This way, devs with already existing frontends like React/Vue/Angular/Ember/etc... can start to trickle in hyperapp and share Redux stores with the native hyperapp emit.

I dunno, either way it is always fun to explore what can be done 😂

**Comment by lukejacksonn** [comment]
@selfup nice one! This is how I saw it being useful 👍

**Comment by jbucaran** [comment]
@dodekeract Implemented in #317.
